lib-src/libcs/getchr.3
.\" COPYRIGHT NOTICE
.\" Copyright (c) 1994 Carnegie Mellon University
.\" All Rights Reserved.
.\" 
.\" See <cmu_copyright.h> for use and distribution information.
.\" 
.\" 
.\" HISTORY
.\" $Log: getchr.3,v $
.\" Revision 4.1  1997/01/08 21:53:52  rvb
.\" r = 4.1; fix $ HEADERS
.\"
.\" Revision 1.1  1996/11/22 19:19:32  braam
.\" First Checkin (pre-release)
.\"
Revision 1.1  96/06/03  19:00:25  satya
Initial revision

.\" Revision 1.2  1995/10/18  14:58:19  moore
.\" 	Created libcs man(3) pages from mach entries
.\" 	[1995/10/18  14:54:06  moore]
.\"
.\" $EndLog$
.\" Copyright (c) 1990 Carnegie Mellon University
.\" All Rights Reserved.
.\" 
.\" Permission to use, copy, modify and distribute this software and its
.\" documentation is hereby granted, provided that both the copyright
.\" notice and this permission notice appear in all copies of the
.\" software, derivative works or modified versions, and any portions
.\" thereof, and that both notices appear in supporting documentation.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND CARNEGIE MELLON UNIVERSITY
.\" DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING ALL
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS.  IN NO EVENT
.\" SHALL CARNEGIE MELLON UNIVERSITY BE LIABLE FOR ANY SPECIAL, DIRECT,
.\" INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER
.\" RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF
.\" CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
.\" CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.\" Users of this software agree to return to Carnegie Mellon any
.\" improvements or extensions that they make and grant Carnegie the
.\" rights to redistribute these changes.
.\"
.\" Export of this software is permitted only after complying with the
.\" regulations of the U.S. Deptartment of Commerce relating to the
.\" Export of Technical Data.
.\"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.\" HISTORY
.\" $Log: getchr.3,v $
.\" Revision 4.1  1997/01/08 21:53:52  rvb
.\" r = 4.1; fix $ HEADERS
.\"
.\" Revision 1.1  1996/11/22 19:19:32  braam
.\" First Checkin (pre-release)
.\"
Revision 1.1  96/06/03  19:00:25  satya
Initial revision

.\" Revision 1.2  1995/10/18  14:58:19  moore
.\" 	Created libcs man(3) pages from mach entries
.\" 	[1995/10/18  14:54:06  moore]
.\"
.\" Revision 1.1.1.2  1995/10/18  14:54:06  moore
.\" 	Created libcs man(3) pages from mach entries
.\"
.\" Revision 1.2  90/12/12  15:50:31  mja
.\" 	Add copyright/disclaimer for distribution.
.\" 
.\" 13-Nov-86  Andi Swimmer (andi) at Carnegie-Mellon University
.\" 	Revised for 4.3.
.\" 
.\" 23-Oct-82  Steven Shafer (sas) at Carnegie-Mellon University
.\" 	Error or EOF on standard input now causes default value to be used.
.\" 
.\" 05-Dec-79  Steven Shafer (sas) at Carnegie-Mellon University
.\" 	Created.
.\" 
.TH GETCHR 3 10/23/82
.CM 3
.SH "NAME"
getchr \- ask user to type a character
.SH "SYNOPSIS"
.B
#include <stdio.h>
.sp
.B
int getchr (prompt,legals,defalt);
.br
.B
char *prompt,*legals,defalt;
.SH "DESCRIPTION"
.I
Getchr
will print its prompt and ask the user for a single character.
If that
charactter is in the string of legal response characters
.RI ( legals ),
then its index is returned.
If a null string is typed (just a
carriage return), then the default character
.RI ( defalt )
will be assumed and its index returned.
If a character is typed
which is not in the string of legal responses, then an error message
is printed and the prompt-and-response cycle is repeated.
.SH "LIBRARY NOTES"
You must link against /usr/local/lib/libcs.a to define this routine.
.SH "SEE ALSO"
.IR getbool (3), 
.IR getstr (3), 
.IR getstab (3), 
etc.
.br
.IR chrarg (3)
.SH "BUGS"
There is no provision for an escape response (e.g.
"I don't like any
of these choices").


NEWS
Incompatible RVM update (Coda client)
-------------------------------------

2008/02/06 RVM version 39
    Move store identifier from per-volume to global rvg struct.

2005/06/07 RVM version 38
    Added realm::realmid (as we can't use &realm_object on 64-bit systems)

2005/06/16 RVM version 37
    Removed fsdb::device and fsobj::inode


lib-src/libcs/getchr.c
#ifndef _BLURB_
#define _BLURB_
/*
 * This code was originally part of the CMU SCS library "libcs".
 * A branch of that code was taken in May 1996, to produce this
 * standalone version of the library for use in Coda and Odyssey
 * distribution.  The copyright and terms of distribution are
 * unchanged, as reproduced below.
 *
 * Copyright (c) 1990-96 Carnegie Mellon University
 * All Rights Reserved.
 * 
 * Permission to use, copy, modify and distribute this software and its
 * documentation is hereby granted, provided that both the copyright
 * notice and this permission notice appear in all copies of the
 * software, derivative works or modified versions, and any portions
 * thereof, and that both notices appear in supporting documentation.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND CARNEGIE MELLON UNIVERSITY
 * DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING ALL
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS.  IN NO EVENT
 * SHALL CARNEGIE MELLON UNIVERSITY BE LIABLE FOR ANY SPECIAL, DIRECT,
 * INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER
 * RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF
 * CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
 * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *
 * Users of this software agree to return to Carnegie Mellon any
 * improvements or extensions that they make and grant Carnegie the
 * rights to redistribute these changes.
 *
 * Export of this software is permitted only after complying with the
 * regulations of the U.S. Deptartment of Commerce relating to the
 * Export of Technical Data.
 */

static char *rcsid = "/afs/cs/project/coda-rvb/cvs/src/coda-4.0.1/lib-src/libcs/getchr.c,v 1.1.1.1 1996/11/22 19:19:49 rvb Exp";
#endif /*_BLURB_*/

/*  getchr  --  ask user to select a character
 *
 *  Usage:  i = getchr (prompt,legals,defalt)
 *	int i;
 *	char *prompt, *legals, defalt;
 *
 *  Prints the message:		prompt  (legals)  [defalt]
 *  and allows the user to type a line.  The first character
 *  the user types is searched for in the string "legals"; if it
 *  is there, its index is returned; otherwise, the user must
 *  try again.  If the user just types carriage return, the
 *  "defalt" character will be searched for.  In any case, note
 *  that is the INDEX (i.e. 0, 1, 2, ...) of the
 *  character that is returned.
 *  On error or EOF in the standard input, the default is used.
 *
 */

#include <stdio.h>
#include <string.h>

#include "libcs.h"

int getchr (prompt, legals, defalt)
const char *prompt, *legals, defalt;
{
	register int i = 0;
	register char *p;
	char input [200];

	fflush (stdout);
	do {
		fprintf (stderr,"%s  (%s)  [%c]  ",prompt,legals,defalt);
		fflush (stderr);
		if (gets (input) == NULL)  *input = defalt;
		if (*input == '\0')  *input = defalt;
		p = strchr (legals, *input);
		if (p == 0)
			fprintf (stderr,"Must be one of: %s\n",legals);
		else
			i = (p - legals);
	} 
	while (p == 0);

	return (i);
}
